File: Sources/ConversationView.swift
import SwiftUI

/// The conversation an interrupt points at — the agent's report_status trail
/// plus the ask itself, answerable inline with quick replies or voice.
struct InterruptConversationView: View {
    @EnvironmentObject var store: AppStore
    @Environment(\.colorScheme) private var scheme
    let interruptId: String

    private var interrupt: Interrupt? {
        store.interrupts.first { $0.id == interruptId }
    }

    var body: some View {
        VStack(spacing: 0) {
            ScrollView {
                VStack(alignment: .leading, spacing: 12) {
                    if let interrupt {
                        contextChip(interrupt)
                    }
                    ForEach(store.thread(for: interruptId)) { msg in
                        bubble(msg)
                    }
                    if interrupt?.kind == .approval && !(interrupt?.resolved ?? true) {
                        approvalCard
                    }
                    if let interrupt, interrupt.resolved {
                        HStack(spacing: 6) {
                            Image(systemName: "checkmark.circle.fill")
                                .font(.system(size: 12))
                                .foregroundStyle(DT.live(scheme))
                            Text("Cleared — \(interrupt.agentName) is moving again")
                                .font(.system(size: 11.5, weight: .semibold))
                                .foregroundStyle(DT.ink55(scheme))
                        }
                        .frame(maxWidth: .infinity)
                        .padding(.top, 6)
                    }
                }
                .padding(.horizontal, 20)
                .padding(.top, 12)
                .padding(.bottom, 16)
            }

            if let interrupt, !interrupt.resolved, interrupt.kind == .blocked {
                quickReplies(["Use env", "Use vault", "Hold on — let's talk"])
            } else if let interrupt, !interrupt.resolved, interrupt.kind == .review {
                quickReplies(["Open the plan", "Park it for today"])
            }

            inputBar
        }
        .background(DT.page(scheme).ignoresSafeArea())
        .navigationTitle(interrupt?.agentName ?? "Conversation")
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .topBarTrailing) {
                Button { store.joinCall() } label: {
                    HStack(spacing: 5) {
                        Image(systemName: "waveform")
                            .font(.system(size: 11, weight: .semibold))
                        Text("Join session")
                            .font(.system(size: 12, weight: .bold))
                    }
                    .foregroundStyle(DT.violetText(scheme))
                    .padding(.horizontal, 10).padding(.vertical, 6)
                    .background(DT.violet.opacity(0.10))
                    .clipShape(Capsule())
                }
                .buttonStyle(Pressable())
            }
        }
    }

    private func contextChip(_ interrupt: Interrupt) -> some View {
        HStack(spacing: 7) {
            Circle().fill(interrupt.resolved ? DT.live(scheme) : DT.violet)
                .frame(width: 6, height: 6)
            Text(interrupt.title)
                .font(.system(size: 11, weight: .semibold))
                .foregroundStyle(DT.ink55(scheme))
                .lineLimit(1)
            Spacer()
            Text(interrupt.age)
                .font(.system(size: 10, design: .monospaced))
                .foregroundStyle(DT.ink35(scheme))
        }
        .padding(.horizontal, 12).padding(.vertical, 8)
        .background(DT.surface2(scheme))
        .clipShape(Capsule())
    }

    @ViewBuilder
    private func bubble(_ msg: ConvMessage) -> some View {
        switch msg.sender {
        case .system:
            HStack(spacing: 6) {
                Image(systemName: "waveform.path.ecg")
                    .font(.system(size: 9))
                Text(msg.text)
                    .font(.system(size: 10.5, design: .monospaced))
                    .lineLimit(1)
                Spacer(minLength: 0)
                Text(msg.time)
                    .font(.system(size: 9.5, design: .monospaced))
            }
            .foregroundStyle(DT.ink35(scheme))
            .padding(.horizontal, 4)
        case .agent:
            HStack(alignment: .bottom, spacing: 9) {
                if let interrupt {
                    AvatarChip(letter: String(interrupt.agentName.prefix(1)),
                               color: interrupt.agentColor, size: 26)
                        .accessibilityHidden(true)
                }
                Text(msg.text)
                    .scaledFont(14)
                    .lineSpacing(2.5)
                    .foregroundStyle(DT.ink(scheme))
                    .padding(.horizontal, 14).padding(.vertical, 11)
                    .background(DT.surface(scheme))
                    .clipShape(RoundedRectangle(cornerRadius: DT.rCard, style: .continuous))
                    .overlay(RoundedRectangle(cornerRadius: DT.rCard, style: .continuous)
                        .strokeBorder(DT.hairline(scheme), lineWidth: 0.8))
                Spacer(minLength: 28)
            }
        case .you:
            HStack(alignment: .bottom) {
                Spacer(minLength: 48)
                Text(msg.text)
                    .scaledFont(14, .medium)
                    .foregroundStyle(.white)
                    .padding(.horizontal, 14).padding(.vertical, 11)
                    .background(DT.heroGradient)
                    .clipShape(RoundedRectangle(cornerRadius: DT.rCard, style: .continuous))
            }
        }
    }

    private var approvalCard: some View {
        VStack(alignment: .leading, spacing: 0) {
            VStack(alignment: .leading, spacing: 4) {
                Text("+ export function requireAuth()")
                Text("+   verifyJwt(req); next()")
            }
            .font(.system(size: 11.5, design: .monospaced))
            .foregroundStyle(scheme == .dark ? DT.live(.dark) : DT.diffGreenOnLight)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 13).padding(.vertical, 11)
            .background(DT.surface2(scheme))
            .clipShape(RoundedRectangle(cornerRadius: DT.rControl, style: .continuous))

            HStack(spacing: 9) {
                Button { store.denyEdit() } label: {
                    Text("Deny")
                        .font(.system(size: 14, weight: .bold))
                        .foregroundStyle(DT.ink78(scheme))
                        .frame(maxWidth: .infinity).frame(height: 44)
                        .background(DT.surface2(scheme))
                        .clipShape(RoundedRectangle(cornerRadius: DT.rControl, style: .continuous))
                }
                .buttonStyle(Pressable())
                Button { store.allowEdit() } label: {
                    Text("Allow")
                        .font(.system(size: 14, weight: .bold))
                        .foregroundStyle(.white)
                        .frame(maxWidth: .infinity).frame(height: 44)
                        .background(DT.heroGradient)
                        .clipShape(RoundedRectangle(cornerRadius: DT.rControl, style: .continuous))
                }
                .buttonStyle(Pressable())
            }
            .padding(.top, 12)
        }
        .padding(14)
        .background(DT.surface(scheme))
        .clipShape(RoundedRectangle(cornerRadius: DT.rCard, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: DT.rCard, style: .continuous)
            .strokeBorder(DT.violet.opacity(0.28), lineWidth: 0.8))
    }

    private func quickReplies(_ options: [String]) -> some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(options, id: \.self) { option in
                    Button { store.sendReply(interruptId: interruptId, option) } label: {
                        Text(option)
                            .font(.system(size: 13, weight: .bold))
                            .foregroundStyle(DT.violetText(scheme))
                            .padding(.horizontal, 14).padding(.vertical, 9)
                            .background(DT.violet.opacity(0.10))
                            .clipShape(Capsule())
                            .overlay(Capsule().strokeBorder(DT.violet.opacity(0.22), lineWidth: 0.8))
                    }
                    .buttonStyle(Pressable())
                }
            }
            .padding(.horizontal, 20)
        }
        .padding(.vertical, 10)
    }

    private var inputBar: some View {
        HStack(spacing: 10) {
            Text("Message \(interrupt?.agentName ?? "agent")…")
                .font(.system(size: 14))
                .foregroundStyle(DT.ink35(scheme))
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, 14)
                .frame(height: 44)
                .background(DT.surface2(scheme))
                .clipShape(RoundedRectangle(cornerRadius: 22, style: .continuous))
            Image(systemName: "mic.fill")
                .font(.system(size: 15, weight: .semibold))
                .foregroundStyle(.white)
                .frame(width: 44, height: 44)
                .background(DT.heroGradient, in: Circle())
                .shadow(color: DT.violet.opacity(0.3), radius: 8, y: 4)
                .accessibilityAddTraits(.isButton)
                .accessibilityLabel("Dictate a reply")
        }
        .padding(.horizontal, 20)
        .padding(.bottom, 10)
    }
}

File: web/src/views/home/ConversationView.js
// ConversationView.swift — the conversation an interrupt points at: the
// agent's report_status trail plus the ask, answerable inline with quick
// replies, a composer, or hold-to-talk. Voice is *measured* (a live level for
// the waveform) and dropped in the same frame — never transcribed or kept.
import { html, useState, useEffect, useRef, useObservable, haptic, clamp } from "../../ui.js";
import { Screen, AvatarChip, Icon, Waveform } from "../../components.js";
import { ctx } from "./shared.js";

const QUICK = { blocked: ["Use env", "Use vault", "Hold on — let's talk"], review: ["Open the plan", "Park it for today"] };

export function ConversationView({ params }) {
  const s = useObservable(ctx.store);
  const nav = ctx.nav;
  const interruptId = params?.interruptId;
  const interrupt = s.interrupts.find((i) => i.id === interruptId) ?? null;
  const thread = s.thread(interruptId);
  const scrollRef = useRef();
  const [draft, setDraft] = useState("");
  const voice = useVoiceLevel();

  // Newest message stays in view — a reply lands, the ack lands 1.2s later.
  useEffect(() => {
    const el = scrollRef.current;
    if (!el) return;
    el.scrollTo({ top: el.scrollHeight, behavior: "auto" });
  }, [thread.length]);

  const send = (text) => {
    const t = String(text ?? "").trim();
    if (!t || !interruptId) return;
    haptic("light");
    s.sendReply(interruptId, t);
    setDraft("");
  };

  const quick = interrupt && !interrupt.resolved ? QUICK[interrupt.kind] : null;
  const name = interrupt?.agentName ?? "agent";

  const footer = html`<div class="cv-footer" style=${{ paddingBottom: s.tabBarVisible ? "calc(var(--tabbar-h) + var(--safe-bottom) + 6px)" : "calc(var(--safe-bottom) + 6px)" }}>
    ${quick ? html`<div class="cv-quick" role="group" aria-label="Quick replies">
      ${quick.map((q) => html`<button key=${q} class="q pressable" onClick=${() => send(q)}>${q}</button>`)}
    </div>` : null}
    <div class="cv-bar">
      ${voice.held ? html`<div class="cv-voice" role="status" aria-live="assertive">
          <${Waveform} color="var(--violet-text)" barCount=${9} height=${20} live=${voice.capturing} level=${voice.level} />
          <span class="truncate">${voice.denied ? "No mic — level only, never audio" : voice.capturing ? "Listening — level only, nothing is kept" : "Hold to talk"}</span>
        </div>`
      : html`<label class="cv-field">
          <input value=${draft} placeholder=${`Message ${name}…`} aria-label=${`Message ${name}`} enterkeyhint="send"
            onInput=${(e) => setDraft(e.target.value)} onKeyDown=${(e) => { if (e.key === "Enter") { e.preventDefault(); send(draft); } }} />
          <button class="send pressable" aria-label="Send" disabled=${!draft.trim()} onClick=${() => send(draft)}><${Icon} name="paperplane.fill" size=${15} weight=${2.4} fill /></button>
        </label>`}
      <button class=${voice.held ? "cv-mic held" : "cv-mic"} aria-label="Hold to talk" aria-pressed=${voice.held} title="Hold to talk — measured, never transcribed"
        onPointerDown=${(e) => { e.preventDefault(); try { e.currentTarget.setPointerCapture(e.pointerId); } catch { /* synthetic pointer */ } voice.start(); }}
        onPointerUp=${voice.stop} onPointerCancel=${voice.stop} onLostPointerCapture=${voice.stop}
        onKeyDown=${(e) => { if ((e.key === " " || e.key === "Enter") && !e.repeat) { e.preventDefault(); voice.start(); } }}
        onKeyUp=${(e) => { if (e.key === " " || e.key === "Enter") voice.stop(); }}
        onContextMenu=${(e) => e.preventDefault()}>
        <${Icon} name="mic.fill" size=${17} weight=${2.4} fill />
      </button>
    </div>
  </div>`;

  return html`<${Screen} title=${interrupt?.agentName ?? "Conversation"} back scrollRef=${scrollRef} contentClass="hm-content" footer=${footer}
    trailing=${html`<button class="hm-wash pressable" onClick=${() => s.joinCall()} aria-label="Join session"><${Icon} name="waveform" size=${12} weight=${2.6} /> Join session</button>`}>
    <div class="cv-thread" data-surface="conversation" role="log" aria-live="polite" aria-label=${`Conversation with ${name}`}>
      ${interrupt ? html`<div class="cv-context">
        <i class=${interrupt.resolved ? "dot" : "dot violet"} aria-hidden="true" />
        <span class="t truncate">${interrupt.title}</span>
        <span class="mono faint" style=${{ fontSize: 10 }}>${interrupt.age}</span>
      </div>` : null}
      ${thread.map((m) => html`<${Bubble} key=${m.id} msg=${m} interrupt=${interrupt} />`)}
      ${interrupt?.kind === "approval" && !interrupt.resolved ? html`<${ApprovalCard} store=${s} />` : null}
      ${interrupt?.resolved ? html`<div class="cv-cleared fade-in">
        <${Icon} name="checkmark.circle.fill" size=${12} fill color="var(--live)" />
        <span>Cleared — ${interrupt.agentName} is moving again</span>
      </div>` : null}
    </div>
  </${Screen}>`;
}

function Bubble({ msg, interrupt }) {
  if (msg.sender === "system") {
    return html`<div class="cv-sys" role="note">
      <${Icon} name="waveform.path.ecg" size=${10} weight=${2.4} />
      <span class="t truncate">${msg.text}</span>
      <span style=${{ fontSize: 9.5 }}>${msg.time}</span>
    </div>`;
  }
  if (msg.sender === "agent") {
    return html`<div class="cv-agent bounce-in">
      ${interrupt ? html`<${AvatarChip} letter=${interrupt.agentName[0]} color=${interrupt.agentColor} size=${26} />` : null}
      <div class="b">${msg.text}</div>
    </div>`;
  }
  return html`<div class="cv-you bounce-in"><div class="b">${msg.text}</div></div>`;
}

function ApprovalCard({ store }) {
  return html`<div class="cv-approval" role="group" aria-label="Approve Cline's edit">
    <div class="ny-diff" style=${{ marginTop: 0 }}><span>+ export function requireAuth()</span><span>+   verifyJwt(req); next()</span></div>
    <div class="ny-actions" style=${{ marginTop: 12 }}>
      <button class="ny-btn ghost pressable" onClick=${() => store.denyEdit()}>Deny</button>
      <button class="ny-btn solid pressable" onClick=${() => store.allowEdit()}>Allow</button>
    </div>
  </div>`;
}

/**
 * VoiceCapture.swift for the browser: getUserMedia → AnalyserNode → RMS → dB
 * → 0…1, smoothed 0.35 like the Swift tap. The graph never reaches a
 * destination, nothing is recorded, and the stream is stopped on release.
 * Without a mic (denied, headless, insecure origin) the waveform idles.
 */
function useVoiceLevel() {
  const [held, setHeld] = useState(false);
  const [capturing, setCapturing] = useState(false);
  const [denied, setDenied] = useState(false);
  const [level, setLevel] = useState(0);
  const ref = useRef({ holding: false, audio: null, stream: null, raf: 0, smooth: 0 });

  const teardown = () => {
    const r = ref.current;
    cancelAnimationFrame(r.raf); r.raf = 0;
    r.stream?.getTracks().forEach((t) => t.stop()); r.stream = null;
    r.audio?.close().catch(() => {}); r.audio = null;
    r.smooth = 0;
  };

  const start = async () => {
    const r = ref.current;
    if (r.holding) return;
    r.holding = true; setHeld(true); haptic("medium");
    if (!navigator.mediaDevices?.getUserMedia || !(window.AudioContext || window.webkitAudioContext)) { setDenied(true); return; }
    let stream;
    try { stream = await navigator.mediaDevices.getUserMedia({ audio: true, video: false }); }
    catch { setDenied(true); return; }
    if (!r.holding) { stream.getTracks().forEach((t) => t.stop()); return; }
    setDenied(false);
    const audio = new (window.AudioContext || window.webkitAudioContext)();
    const analyser = audio.createAnalyser();
    analyser.fftSize = 1024;
    audio.createMediaStreamSource(stream).connect(analyser);
    r.stream = stream; r.audio = audio;
    const buf = new Float32Array(analyser.fftSize);
    const loop = () => {
      analyser.getFloatTimeDomainData(buf);
      let sum = 0;
      for (let i = 0; i < buf.length; i++) sum += buf[i] * buf[i];
      const rms = Math.sqrt(sum / buf.length);
      const db = 20 * Math.log10(Math.max(rms, 1e-7));
      const normalized = clamp((db + 50) / 50, 0, 1);
      r.smooth += (normalized - r.smooth) * 0.35;
      setLevel(r.smooth);
      r.raf = requestAnimationFrame(loop);
    };
    setCapturing(true);
    r.raf = requestAnimationFrame(loop);
  };

  const stop = () => {
    const r = ref.current;
    if (!r.holding) return;
    r.holding = false;
    teardown();
    setHeld(false); setCapturing(false); setLevel(0);
  };

  useEffect(() => () => { ref.current.holding = false; teardown(); }, []);
  return { held, capturing, denied, level, start, stop };
}
